Print errors at the bottom when using Spec reporter

picandocodigo opened the pull request:
This is an initial attempt for a modification to the Spec reporter that is much more convenient for me personally. I'm using this reporter in [`elasticsearch-xpack`](https://github.com/elastic/elasticsearch-ruby/tree/master/elasticsearch-xpack), and it would be much more practical for me to see the errors at the bottom. With this change, the error is still reported with the rest of the tests:
![image](https://user-images.githubusercontent.com/689327/101149137-6bc22c80-3616-11eb-82d7-dcf3f3ac0671.png)

But you can see the detail in errors or failures at the bottom of the report:
![image](https://user-images.githubusercontent.com/689327/101054093-6d411580-3580-11eb-92d4-ea60b2036c9a.png)

Would this be convenient to add? Maybe as an option?

Related: #304

mtomov commented:
Hey @picandocodigo , thanks for your PR - we've started using your fork as indeed errors at the bottom is more practical.

Thank you!

picandocodigo commented:
@mtomov That's great to know! :+1: :tada: 

I'm hesitant to publish the fork in Rubygems yet, since I'd rather wait for this project to find its maintainers. Hopefully we'll get a maintainer soon :crossed_fingers:

itsliamegan commented:
@picandocodigo We've indeed found new maintainers: myself and @barrettkingram 😄 Thanks for the PR; this looks like a great and helpful change! I have one concern, though.

In the "failures and errors" section, the name of the failing test only shows the name of the method, not the class. e.g. if I have a failing test `MyLibrary::MyTest#test_method`, the failure will only show `test_method`. Seeing as you and @mtomov have actually used this reporter for real work, have you found that to be a hinderance when debugging test failures?

Either way, I think this is good to merge. I don't think showing the fully qualified test name is a dealbreaker, just something to keep in mind.

os97673 commented:
I'd also suggest to add a flag to control this behavior (disabled for backward compatibility or not it is up to you)

picandocodigo commented:
Thanks for the feedback @itsliamegan and @os97673!

I'll take a look at showing the full test name and putting this behind a flag as soon as possible and update the Pull Request.

picandocodigo commented:
Hi @itsliamegan @barrettkingram and @os97673 :wave: 
Sorry that it's been a while, I've updated this Pull Request. I made it an option in the reporter's initializer and added some documentation to it. By default it's going to behave the same, but you can set the option to print the errors at the bottom with:

```ruby
Minitest::Reporters.use! Minitest::Reporters::SpecReporter.new(print_at_bottom: true)
```
I also added the class name to the report like @itsliamegan suggested, I think it's useful information that was missing in my original commit.

Let me know what you think and if any more changes are needed please.
Thanks!

picandocodigo commented on lib/minitest/reporters/spec_reporter.rb:
```diff
@@ -10,6 +10,16 @@ class SpecReporter < BaseReporter
+      #
+      def initialize(options = {})
+        super
+        @print_at_bottom = options[:print_at_bottom]
```
I'm not convinced about the name for the option here, so suggestions are very welcome :smile:

picandocodigo commented:
Hi @itsliamegan @barrettkingram and @os97673, wondering if the project is still being maintained or it's worth maintaining a fork?
Thanks!

os97673 commented:
> Hi @itsliamegan @barrettkingram and @os97673, wondering if the project is still being maintained or it's worth maintaining a fork?
> Thanks!

Hi @picandocodigo ,

I’ve transferred maintenance of the project to @itsliamegan and @barrettkingram so it mostly question for them. As for maintaining a fork it depends 😉 if you are interested in maintaining the library it is possible to add you to the organization of the feature is the only thing you are passionate about then you have to maintain the fork until the PR will be merged 😒

picandocodigo commented:
Hey @os97673,
Thanks for your reply. I can help maintain the library if @itsliamegan and @barrettkingram are ok with it :+1:

barrettkingram commented on lib/minitest/reporters/spec_reporter.rb:
```diff
@@ -18,6 +28,14 @@ def start
       def report
         super
+        if @print_at_bottom
+          failed_tests = tests.reject { |test| test.failures.empty? }
```
```suggestion
          failed_test_groups = tests
                               .reject { |test| test.failures.empty? }
                               .sort_by { |test| [test_class(test).to_s, test.name] }
                               .group_by { |test| test_class(test).to_s }
```
I think we can make the output a little cleaner by sorting and grouping the test failures by class name. Then we could print the class name just once per group. For example, the current output generated by the branch in this PR might look something like this:
```
Failures and errors:
MinitestReportersTest::BadTest
  test_a                                                          FAIL (0.00s)
        Expected: 1
          Actual: 2
        /Users/bingram/Projects/minitest-reporters/test/gallery/bad_test.rb:6:in `test_a'
Location:
         /Users/bingram/Projects/minitest-reporters/test/gallery/bad_test.rb:5


MinitestReportersTest::BadTest
  test_b                                                          FAIL (0.00s)
        Expected false to be truthy.
        /Users/bingram/Projects/minitest-reporters/test/gallery/bad_test.rb:10:in `test_b'
Location:
         /Users/bingram/Projects/minitest-reporters/test/gallery/bad_test.rb:9
```
With results grouped by test class, we could print something like this instead:
```
Failures and errors:
MinitestReportersTest::BadTest
  test_a                                                          FAIL (0.00s)
        Expected: 1
          Actual: 2
        /Users/bingram/Projects/minitest-reporters/test/gallery/bad_test.rb:6:in `test_a'
Location:
         /Users/bingram/Projects/minitest-reporters/test/gallery/bad_test.rb:5

  test_b                                                          FAIL (0.00s)
        Expected false to be truthy.
        /Users/bingram/Projects/minitest-reporters/test/gallery/bad_test.rb:10:in `test_b'
Location:
         /Users/bingram/Projects/minitest-reporters/test/gallery/bad_test.rb:9
```

picandocodigo replied:
I updated the code to sort and group the errors, makes sense, it looks much cleaner this way, thanks! :+1:

barrettkingram commented on lib/minitest/reporters/spec_reporter.rb:
```diff
@@ -42,11 +60,12 @@ def after_suite(_suite)
+        puts test.klass
+        record_print_status(test)
+        print_info(test.failure, test.error?)
+        puts "Location:\n\t #{test.source_location.join(':')}"
```
Hmm what value does printing the `source_location` add since we already print the backtrace?

barrettkingram commented:
@picandocodigo Apologies for the delay. Left a couple suggestions. Thanks for working on this - this is a fantastic improvement!

picandocodigo commented:
Thanks @os97673 for the response and @barrettkingram for the feedback!

barrettkingram commented:
@picandocodigo thanks for addressing feedback! This is looking really good. What do you think about my suggestion for changing the option name to `print_failure_summary`? Once that is addressed this should be ready to merge.

picandocodigo commented:
@barrettkingram updated :+1:

barrettkingram commented:
Thanks, will put together a release soon!

picandocodigo commented:
@barrettkingram sorry to bother, but are there plans for a release soon?

barrettkingram commented:
@picandocodigo I've just pushed 1.6.0 which includes this PR. Thanks for your patience!
